Knockback v2

coolpeter98 opened the pull request:
Various improvements to knockback and combat.. this pretty much doesn't use any of Eli's knockback code.

coolpeter98 committed:
Various combat and codebase improvements.

coolpeter98 committed:
Knockback improvements

coolpeter98 committed:
Remove outdated comment

coolpeter98 committed:
Remove useless script in ServerScriptService

coolpeter98 committed:
Use server-sided knockback instead of sending remote events

EliTheGingerCat commented on src/ReplicatedStorage/modules/sword_cooldown.luau:
```diff
@@ -0,0 +1,3 @@
+sword_cooldown = 0.5
+
```
`local sword_cooldown...`

EliTheGingerCat commented on src/ReplicatedStorage/modules/wait_for_frames.luau:
```diff
@@ -0,0 +1,10 @@
+local RunService = game:GetService("RunService")
+
+local function wait_for_frames(frames)
```
`frames: number?`
function return: `()`

EliTheGingerCat commented on src/ServerScriptService/modules/sword_cooldown_server.luau:
```diff
@@ -0,0 +1,7 @@
+
+local sword_cooldown = require(ReplicatedStorage.modules.sword_cooldown)
+
+sword_cooldown_server = sword_cooldown - 0.1 --why? to compensate for ping.
```
`local sword_cooldown_server...`

EliTheGingerCat commented on src/ReplicatedStorage/modules/is_on_cooldown.luau:
```diff
@@ -0,0 +1,11 @@
+	if not playerCooldowns then return false end
+
+	local endTime = playerCooldowns[name]
+	return endTime ~= nil and endTime > os.clock()
```
Why would `endTime` be nil?

Also, please use `snake_case`.

EliTheGingerCat commented on src/ServerScriptService/add_hitbox.server.luau:
```diff
@@ -17,7 +17,7 @@ local function on_character_added(character: Model): ()
 	weld.Part1 = root
 
-	clone.Parent = character
+	clone.Parent = character -- changed to parent to root instead because parenting to character made roblox goof with the size for some reason..
```
The comment does not make sense. This code snippet still parents to the character, not the humanoid root part.

EliTheGingerCat commented on src/ServerScriptService/test_attacking.server.luau:
As a note, this is taken from <https://github.com/EliTheGingerCat/libre-duel/pull/34>, specifically <https://github.com/EliTheGingerCat/libre-duel/pull/34/commits/b5640a716f210aeaaf24fdcae0fe5c8d4006ee04#diff-0ba15caae3981b2e1dd4dc20003443fa07393e43805d03b05d3a45aa686cdd53>.

EliTheGingerCat commented on src/ServerStorage/hitbox.rbxm:
Not sure why this file was edited but probably some minor encoding issue that affects nothing.

EliTheGingerCat commented on src/ServerScriptService/attack_player.server.luau:
```diff
@@ -1,9 +1,15 @@
 
-local DAMAGE = 10
+local sword_damage = require(ReplicatedStorage.modules.sword_damage)
+local sword_cooldown_server = require(script.Parent.modules.sword_cooldown_server)
```
I prefer all requires together, though the Roblox Style Guide _does_ recommend blocks of requires.

See: <https://roblox.github.io/lua-style-guide/#requires-structure>

EliTheGingerCat commented on src/ServerScriptService/attack_player.server.luau:
```diff
@@ -42,10 +48,19 @@ local function on_event(player: Player, target_character: unknown): ()
+		local attack_vector = target_position - attacker_position
 
-	play_sound_excluding(player, "player_damaged", attacker_position)
+		local target_player = game.Players:GetPlayerFromCharacter(target_character)
```
Please import `Players` at the top.

EliTheGingerCat replied:
This line of code was entirely removed.

EliTheGingerCat commented on src/ServerScriptService/attack_player.server.luau:
```diff
@@ -42,10 +48,19 @@ local function on_event(player: Player, target_character: unknown): ()
+		local target_player = game.Players:GetPlayerFromCharacter(target_character)
+		apply_knockback(attacker_root, target_root, 1.2, true)
+
+		play_sound_excluding(player, "player_damaged", target_position)
```
Nice catch that we must use `target_position`.

EliTheGingerCat commented on src/ServerScriptService/attack_player.server.luau:
```diff
@@ -42,10 +48,19 @@ local function on_event(player: Player, target_character: unknown): ()
 	if target_humanoid == nil then
 		return
 	end
+	if attacker_character.Humanoid.Health > 0 and target_humanoid.Health > 0 and is_on_cooldown(player, "attack_melee") == false then --me when the basic sanity checks
```
Every other `if` statement checks for an unwanted situation and returns early if that situation arises. I do not see why this one should be an exception.

EliTheGingerCat replied:
This is a long condition — I would prefer a separate `if` statement for each condition, similar to earlier in this function.

EliTheGingerCat replied:
`attacker_character.Humanoid` can error! Please follow the same procedure as `local target_humanoid = target_character:FindFirstChildOfClass("Humanoid")`.

EliTheGingerCat commented on src/ServerScriptService/attack_player.server.luau:
```diff
@@ -42,10 +48,19 @@ local function on_event(player: Player, target_character: unknown): ()
+		local attack_vector = target_position - attacker_position
 
-	play_sound_excluding(player, "player_damaged", attacker_position)
+		local target_player = game.Players:GetPlayerFromCharacter(target_character)
```
`attack_vector` and `target_player` are unused.

EliTheGingerCat commented on src/ReplicatedStorage/modules/apply_knockback.luau:
```diff
@@ -0,0 +1,31 @@
+		end
+	end
+
+	print("Direction:", direction)
```
Debug print left in.

EliTheGingerCat commented on src/ReplicatedStorage/modules/apply_knockback.luau:
```diff
@@ -0,0 +1,31 @@
+	local attachment = Instance.new("Attachment")
+	attachment.Parent = target
+
+	local linearvelocity = Instance.new("LinearVelocity")
```
Please use `snake_case`.

EliTheGingerCat commented on src/ReplicatedStorage/modules/apply_knockback.luau:
```diff
@@ -0,0 +1,31 @@
+	print("Direction:", direction)
+
+	local attachment = Instance.new("Attachment")
+	attachment.Parent = target
```
Please parent the attachment after it is completely configured.

EliTheGingerCat commented on src/ReplicatedStorage/client/modules/attack_melee.luau:
```diff
@@ -1,6 +1,10 @@
 
+local add_cooldown = require(ReplicatedStorage.modules.add_cooldown)
+local is_on_cooldown = require(ReplicatedStorage.modules.is_on_cooldown)
+local sword_cooldown  = require(ReplicatedStorage.modules.sword_cooldown)
+
 local debug_print = require(script.Parent.debug_print)
 local get_player_position = require(ReplicatedStorage.modules.get_player_position)
 local play_sound_local = require(script.Parent.play_sound_local)
```
Again, I prefer all requires together.

EliTheGingerCat commented on src/ReplicatedStorage/client/modules/attack_melee.luau:
```diff
@@ -40,18 +44,23 @@ local function attack_melee(): ()
 	end
 
 	local target_position = target_root.CFrame.Position
-	
+
```
Nice catch. Did you use a tool for this, or manually spotted?

EliTheGingerCat commented on src/ReplicatedStorage/modules/remove_cooldown.luau:
```diff
@@ -0,0 +1,9 @@
+local cooldowns = require(script.Parent.cooldowns)
+
+local function remove_cooldown(player: Player, name: string)
```
function return: `()`

EliTheGingerCat commented on src/ReplicatedStorage/modules/apply_knockback.luau:
```diff
@@ -0,0 +1,31 @@
+local delay_for_frames = require(script.Parent.delay_for_frames)
+
```
`delay_for_frames` is unused.

EliTheGingerCat commented on src/ReplicatedStorage/modules/apply_knockback.luau:
```diff
@@ -0,0 +1,31 @@
+local delay_for_frames = require(script.Parent.delay_for_frames)
+
+local debris = game:GetService("Debris")
```
Please spell service names exactly as they are named, so using `PascalCase`. Thus: `local Debris...`

EliTheGingerCat commented on src/ReplicatedStorage/modules/apply_knockback.luau:
```diff
@@ -0,0 +1,31 @@
+
+local debris = game:GetService("Debris")
+
+local function apply_knockback(vector, target, strength: number?, constant_strength: boolean?) --constant_strength keeps knockback consistent no matter how far the player is hit from! you wouldn't really want that set to false unless the knockback is from an explosive.
```
This comment should be above the function.

EliTheGingerCat replied:
`vector` should be `attacker`.

EliTheGingerCat commented on src/ReplicatedStorage/modules/apply_knockback.luau:
```diff
@@ -0,0 +1,31 @@
+
+local debris = game:GetService("Debris")
+
+local function apply_knockback(vector, target, strength: number?, constant_strength: boolean?) --constant_strength keeps knockback consistent no matter how far the player is hit from! you wouldn't really want that set to false unless the knockback is from an explosive.
```
`vector: Part, target: Part`
function return: `()`

EliTheGingerCat replied:
The function return type is still not there.

EliTheGingerCat commented on src/ReplicatedStorage/modules/apply_knockback.luau:
```diff
@@ -0,0 +1,31 @@
+	linearvelocity.Attachment0 = attachment
+	linearvelocity.Parent = attachment
+
+	linearvelocity.VectorVelocity = Vector3.new(direction.X, 15, direction.Z) * strength
```
Unfortunately, `strength` is `number?` resulting in a type error. I think it is best to use `local strength = strength or 1` above, even though that results in a [LocalShadow](<https://luau.org/lint#localshadow-4>) error. I think the LocalShadow is better; in fact, I might silence it completely in `.luaurc`. I am not sure but I think the new type solving engine (which is available for use) might address this issue.

EliTheGingerCat reviewed (changes requested):
Thank you! This is really excellent.

One other note is that there is too much modularisation here. I think there is no reason for `wait_for_frames` to be separate from `delay_for_frames`. In theory, there is a case where one would want to wait without creating a new thread for this job, but I prefer crossing such a bridge when we get there. (If this module already existed due to need, and then its dependent was removed, then it would be fine to keep it as is, even though there would no longer be a reason to separate the two modules.)

It is questionable if all the cooldown modules should be separate, but that is fine for now.

I now see the issue of a separate file for every constant, but it is fine for now.

EliTheGingerCat commented:
This pull request is an alternative to <https://github.com/EliTheGingerCat/libre-duel/pull/34>.

coolpeter98 committed:
Remove wait_for_frames modulescript

coolpeter98 committed:
Fix other issues

coolpeter98 committed:
Slightly decrease vertical hitbox side

coolpeter98 commented:
Please review the changes, all issues should have been fixed.

EliTheGingerCat commented:
> Please review the changes, all issues should have been fixed.

By the way, you can re-request a review.

EliTheGingerCat commented on src/ReplicatedStorage/modules/delay_for_frames.luau:
```diff
@@ -0,0 +1,11 @@
+local function delay_for_frames(frames: number, callback: () -> ())
+	task.spawn(function()
+		frames = frames or 1
```
Either frames is optional (`frames: number?`), or there is no need to add a default of `1`.

EliTheGingerCat commented on src/ReplicatedStorage/modules/delay_for_frames.luau:
```diff
@@ -0,0 +1,11 @@
+	task.spawn(function()
+		frames = frames or 1
+		for _ = 1, frames do
+			RunService.Heartbeat:Wait()
```
Erm. Bro. Please use the [Luau Language Server](<https://github.com/JohnnyMorganz/luau-lsp>). `RunService` is not defined in this script.

coolpeter98 replied:
I'm using the Kate text editor because visual studio code is bloat.

EliTheGingerCat commented on src/ReplicatedStorage/modules/is_on_cooldown.luau:
```diff
@@ -0,0 +1,11 @@
+	if not player_cooldowns then return false end
+
+	local end_time = player_cooldowns[name]
+	return end_time > os.clock()
```
I was wrong. `end_time` *can* be `nil`. I would add a separate if statement to return `false` if `end_time == nil`.

Comment where I was wrong: <https://github.com/EliTheGingerCat/libre-duel/pull/35/files/e353e3e302b10e05af58d871c985fe05a174dc9c#r2453666323>

EliTheGingerCat commented on src/ReplicatedStorage/modules/delay_for_frames.luau:
```diff
@@ -0,0 +1,11 @@
+local function delay_for_frames(frames: number, callback: () -> ())
+	task.spawn(function()
```
function return: `()`

EliTheGingerCat reviewed (changes requested):
Seems like not all of my requested changes were added.

EliTheGingerCat committed:
fix: fix various issues

* Keep requires together, ordered descending alphabetically.

* Add and fix function types.

* Remove local shadowing.

See: <https://luau.org/lint#localshadow-4>

* Add missing variables.

* Normalise code style.

* Add check for nil time.

* Remove unused variables.

EliTheGingerCat committed:
feat: assert cooldown names are valid

EliTheGingerCat committed:
feat: add a type and cleanup for cooldowns

EliTheGingerCat committed:
refactor: reorder early returns

It is probably slightly more performant now, but not major enough for
this to be a "perf" commit.

This commit will also:

* Keep requires together, sorted descending alphabetically.

EliTheGingerCat committed:
fix: make healing command work for everyone

EliTheGingerCat committed:
fix: remove debug print

EliTheGingerCat committed:
feat: remove cooldown time when cooled

EliTheGingerCat committed:
feat: improve cooldown type safety

EliTheGingerCat committed:
style: rename variable

EliTheGingerCat committed:
refactor: configure completely before parenting

EliTheGingerCat committed:
docs: add comment author and further explanation

EliTheGingerCat committed:
style: make condition explicit

EliTheGingerCat committed:
refactor: avoid anonymous function

EliTheGingerCat committed:
refactor: redefine variable for a better type

EliTheGingerCat committed:
docs: improve code comment and remove unneeded one

EliTheGingerCat committed:
docs: add knockback to changelog
